packages/extensions/src/horizontal-rule/index.ts

```diff
@@ -0,0 +1,155 @@
+} from '@prosekit/core'
+import { InputRule } from '@prosekit/pm/inputrules'
+import type { Attrs } from '@prosekit/pm/model'
+import {
+  type EditorState,
+  TextSelection,
+  type Transaction,
+} from '@prosekit/pm/state'
+
+import { defineEnterRule } from '../enter-rule'
+import { defineInputRule } from '../input-rule'
+
+function updateFromNodeSelection(
+  tr: Transaction,
+  insertionNode?: string | false,
+): void {
+  // Make sure  the `horizontalRule` that is selected. Otherwise do nothing.
+  if (
+    !isNodeSelection(tr.selection) ||
+    tr.selection.node.type.name !== 'horizontalRule'
+  )
+    return
+
+  if (!insertionNode) {
+    return
+  }
+
+  const pos = tr.selection.$from.pos + 1
+  const type = getNodeType(tr.doc.type.schema, insertionNode)
+
+  const node = type.create()
+  tr.insert(pos, node)
+
+  tr.setSelection(TextSelection.near(tr.doc.resolve(pos + 1)))
+}
+
+type HorizontalRuleSpecExtension = Extension<{
+  Nodes: {
+    horizontalRule: Attrs
+  }
+}>
+
+function defineHorizontalRuleSpec(): HorizontalRuleSpecExtension {
+  return defineNodeSpec({
+    name: 'horizontalRule',
+    group: 'block',
+    parseDOM: [{ tag: 'hr' }],
+    toDOM: () => ['hr'],
+  })
+}
+
+interface InsertHorizontalRuleOptions {
+  insertionNode?: string | false
+}
+
+/**
+ * Adds input rules for `horizontalRule` nodes.
+ *
+ * @public
+ */
+function defineHorizontalRuleInputRule(
+  options?: InsertHorizontalRuleOptions,
+): PlainExtension {
+  function createInsertHorizontalRuleTr(
+    state: EditorState,
+    match: RegExpMatchArray,
+    start: number,
+    end: number,
+  ): Transaction {
+    const { tr, schema } = state
+    const hr = getNodeType(schema, 'horizontalRule')
+    if (match[0]) tr.replaceWith(start - 1, end, hr.create())
+
+    return tr
+  }
+
+  return union(
+    defineInputRule(
+      new InputRule(/^(?:---|___\s|\*\*\*\s)$/, (state, match, start, end) => {
+        const { insertionNode = 'paragraph' } = options || {}
+
+        const tr = createInsertHorizontalRuleTr(state, match, start, end)
+        updateFromNodeSelection(tr, insertionNode)
+
+        return tr
+      }),
+    ),
+    defineEnterRule({
+      regex: /^(?:_{3}|\*{3})$/,
+      handler: ({ state, match, from, to }) => {
+        return createInsertHorizontalRuleTr(state, match, from, to)
+      },
+    }),
+  )
+}
+
+type HorizontalRuleCommandsExtension = Extension<{
+  Commands: {
+    insertHorizontalRule: [options?: InsertHorizontalRuleOptions]
+  }
+}>
+
+function defineHorizontalRuleCommands(): HorizontalRuleCommandsExtension {
+  return defineCommands({
+    insertHorizontalRule: (options?: InsertHorizontalRuleOptions) => {
+      return (state, dispatch) => {
+        if (!dispatch) return true
+        const { tr, schema } = state
+        const hr = getNodeType(schema, 'horizontalRule')
+
+        const _tr = tr.replaceSelectionWith(hr.create())
+        updateFromNodeSelection(_tr, options?.insertionNode ?? 'paragraph')
```

Comment:
I prefer not hard encoding `paragraph`. You can use `insertDefaultBlock()`, which will pick `paragraph` in most cases, but also supports schema without `paragraph` in it.